### PR TITLE
refactor: broadcast availability status using message sender WBP-5802

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Broadcasting the availability status was still using an old non-async code path which was affected by the proteus session methods turning async. Since It would be wasted effort (and complicated) to update this code path to be async I'm deleting by implementing broadcast support in the `MessageSender`.

#### Test Coverage

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/APIs/MessageAPI.swift [hunk 1]
```diff
@@ -58,6 +60,41 @@ class MessageAPIV0: MessageAPI {
...
+
+        let response = await httpClient.send(request)
+
+        if response.httpStatus == 412 {
+            guard
+                let messageSendingStatus = Payload.MessageSendingStatusV0(response, decoder: .defaultDecoder)
+            else {
+                throw NetworkError.errorDecodingResponse(response)
+            }
+            throw NetworkError.missingClients(messageSendingStatus.toMessageSendingStatus(domain: ""), response)
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/APIs/MessageAPI.swift:91]: do you need a domain here like `BackendAPI.domain` ?

File: wire-ios-request-strategy/Sources/Message Sending/MessageSenderTests.swift [hunk 2]
```diff
@@ -117,8 +121,88 @@ final class MessageSenderTests: MessagingTestBase {
...
+        // when
+        try await messageSender.broadcastMessage(message: message)
+
+        // then
+        XCTAssertEqual(2, arrangement.messageApi.broadcastProteusMessageMessage_Invocations.count)
+    }
+
+    func testThatWhenSendingProteusMessageSucceeds_thenCompleteWithoutErrors() async throws {
+        // given
+        let response = ZMTransportResponse(payload: nil, httpStatus: 200, transportSessionError: nil, apiVersion: 0)
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Message Sending/MessageSenderTests.swift:192]: does it make sense to test for other apiVersion?

File: wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift [hunk 3]
```diff
@@ -18,147 +18,46 @@
...
         let recipients = ZMUser.recipientsForAvailabilityStatusBroadcast(in: context, maxCount: maximumBroadcastRecipients)
+        let proteusMessage = GenericMessageEntity(message: message, context: context, targetRecipients: .users(recipients), completionHandler: nil)
 
-        guard let dataAndMissingClientStrategy = message.encryptForTransport(forBroadcastRecipients: recipients, in: context) else {
-            return nil
+        WaitingGroupTask(context: context) { [self] in
+            try? await messageSender.broadcastMessage(message: proteusMessage)
+            await context.perform { [self] in
+                completion()
+                context.enqueueDelayedSave()
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift:58]: normal to call enqueue after completion?

File: wire-ios-request-strategy/Sources/Request Strategies/Delivery Receipts/DeliveryReceiptRequestStrategy.swift [hunk 4]
```diff
@@ -81,8 +81,9 @@ extension DeliveryReceiptRequestStrategy: ZMEventConsumer {
...
         let groups = managedObjectContext.enterAllGroupsExceptSecondary()
         Task {
-            try? await messageSender.sendMessage(message: GenericMessageEntity(conversation: deliveryReceipt.conversation,
-                                                                               message: GenericMessage(content: confirmation),
+            try? await messageSender.sendMessage(message: GenericMessageEntity(message: GenericMessage(content: confirmation),
+                                                                               context: managedObjectContext,
+                                                                               conversation: deliveryReceipt.conversation,
                                                                                targetRecipients: .users(senderUserSet),
                                                                                completionHandler: nil))
             managedObjectContext.leaveAllGroups(groups)
```
Comment (netbe) [inline wire-ios-request-strategy/Sources/Request Strategies/Delivery Receipts/DeliveryReceiptRequestStrategy.swift]: use  WaitingGroupTask(context: managedObjectContext)

File: wire-ios/Wire-iOS.xcodeproj/xcshareddata/xcschemes/Wire-iOS.xcscheme [hunk 5]
```diff
@@ -226,7 +226,7 @@
          </CommandLineArgument>
          <CommandLineArgument
             argument = "-BackendEnvironmentTypeOverrideKey staging"
-            isEnabled = "NO">
+            isEnabled = "YES">
```
Comment (netbe) [inline wire-ios/Wire-iOS.xcodeproj/xcshareddata/xcschemes/Wire-iOS.xcscheme]: to revert

**Review by netbe — COMMENTED** [review]
I am not familiar at all with this broadcasting, so hard to follow.

Added some suggestions about the code though

File: wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift (hunk 3 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift:58]: Inside the completion the `modifiedKeys` of the user are reset we need to save the context after that. Normally this save happen implicitly after http response processing is complete, but since this is happening async we need to manually trigger a save.

File: wire-ios-request-strategy/Sources/Message Sending/MessageSenderTests.swift (hunk 2 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Message Sending/MessageSenderTests.swift:192]: No, it's the same logic

File: wire-ios-request-strategy/Sources/APIs/MessageAPI.swift (hunk 1 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/APIs/MessageAPI.swift:91]: Is `BackendAPI.domain` always set?

Comment (typfel) [reply wire-ios-request-strategy/Sources/APIs/MessageAPI.swift:91]: Last I time looked this would not be necessary, but if I can use `BackendAPI.domain` it would make the code easier to follow.

Anyway I'd prefer to not this change in this PR.

File: wire-ios/Wire-iOS.xcodeproj/xcshareddata/xcschemes/Wire-iOS.xcscheme (hunk 5 above)
Comment (typfel) [reply wire-ios/Wire-iOS.xcodeproj/xcshareddata/xcschemes/Wire-iOS.xcscheme]: Done

File: wire-ios-request-strategy/Sources/Request Strategies/Delivery Receipts/DeliveryReceiptRequestStrategy.swift (hunk 4 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Request Strategies/Delivery Receipts/DeliveryReceiptRequestStrategy.swift]: Done

File: wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift (hunk 3 above)
Comment (caldrian) [reply wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift:58]: Could you put this explanation into a comment at this place, please?
I also tend to make any completion call the last statement in the block when I encounter something like this.

File: wire-ios-request-strategy/Sources/APIs/MessageAPI.swift (hunk 1 above)
Comment (netbe) [reply wire-ios-request-strategy/Sources/APIs/MessageAPI.swift:91]: ok, not always set as it is optional

Comment (typfel) [reply wire-ios-request-strategy/Sources/APIs/MessageAPI.swift:91]: We should check if can't simply delete API v0 support. AFAIK no one is running a backend which only supports V0

File: wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift (hunk 3 above)
Comment (typfel) [reply wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift:58]: Done

**Review by netbe — APPROVED** [review]

**Review by caldrian — APPROVED** [review]
